eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/permission/fiveway/FiveWayPermissionTracker.java

```diff
@@ -0,0 +1,246 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.schedjoules.eventdiscovery.framework.permission.fiveway;
+
+import android.app.Activity;
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.content.pm.PackageInfo;
+import android.content.pm.PackageManager;
+import android.support.annotation.NonNull;
+import android.support.v4.app.ActivityCompat;
+import android.support.v4.content.ContextCompat;
+import android.util.Log;
+
+import com.schedjoules.eventdiscovery.framework.permission.broadcast.PermissionRequestResult;
+import com.schedjoules.eventdiscovery.framework.permission.tracker.PermissionTracker;
+import com.schedjoules.eventdiscovery.framework.permission.tracker.TrackablePermissionRequestResult;
+
+import static android.content.pm.PackageManager.PERMISSION_GRANTED;
+import static com.schedjoules.eventdiscovery.framework.permission.fiveway.FiveWayPermissionStatus.DENIED;
+import static com.schedjoules.eventdiscovery.framework.permission.fiveway.FiveWayPermissionStatus.DENIED_WITH_NEVER_ASK_AGAIN;
+import static com.schedjoules.eventdiscovery.framework.permission.fiveway.FiveWayPermissionStatus.GRANTED;
+import static com.schedjoules.eventdiscovery.framework.permission.fiveway.FiveWayPermissionStatus.NOT_ASKED_YET;
+import static com.schedjoules.eventdiscovery.framework.permission.fiveway.FiveWayPermissionStatus.NOT_IN_MANIFEST;
+
+
+/**
+ * {@link PermissionTracker} for the {@link FiveWayPermissionStatus}s. It uses {@link SharedPreferences} to store the status, it's needed to be able to tell
+ * whether user has checked the "Never ask again" checkbox previously.
+ *
+ * @author Gabor Keszthelyi
+ */
+final class FiveWayPermissionTracker implements PermissionTracker<FiveWayPermissionStatus>
+{
+    private static final String TAG = "PermissionStatus";
+    private static final String SHARED_PREF_STORE_NAME = "SchedJoules.Permissions";
+    private static final boolean LOG_ENABLED = false;
+
+    private static FiveWayPermissionTracker sInstance;
+
+
+    public static FiveWayPermissionTracker getInstance()
+    {
+        if (sInstance == null)
+        {
+            sInstance = new FiveWayPermissionTracker();
+        }
+        return sInstance;
+    }
+
+
+    private FiveWayPermissionTracker()
+    {
+    }
+
+
+    @Override
+    public FiveWayPermissionStatus checkStatus(Activity activity, String permissionName)
+    {
+        return checkStatus(activity, permissionName, true, LOG_ENABLED);
+    }
+
+
+    @Override
+    public TrackablePermissionRequestResult trackableResult(final PermissionRequestResult result)
+    {
+        return new TrackablePermissionRequestResult()
```

Comment:
In `Recent<T> Recents.recent(T value)` I'm using this pattern because the actual `Recent` implementation depends on `Recents` and only `Recents` can create the appropriate class.

I can't see any reason for the same pattern in FiveWayPermissionTracker.